envisage/plugins/ipython_kernel/internal_ipkernel.py

```diff
@@ -3,19 +3,52 @@
 https://github.com/ipython/ipython/blob/2.x/examples/Embedding/internal_ipkernel.py
 
 """
+import atexit
 from distutils.version import StrictVersion as Version
 import sys
 
 import ipykernel
 from ipykernel.connect import connect_qtconsole
-from ipykernel.kernelapp import IPKernelApp
+from ipykernel.ipkernel import IPythonKernel
+from ipykernel.zmqshell import ZMQInteractiveShell
+import IPython.utils.io
+import six
 from tornado import ioloop
 
+from envisage.plugins.ipython_kernel.kernelapp import IPKernelApp
 from traits.api import Any, HasStrictTraits, Instance, List
 
 NEEDS_IOLOOP_PATCH = Version(ipykernel.__version__) >= Version('4.7.0')
 
 
+# XXX Refactor to avoid the fragile subclassing of IPKernelApp. (One day.)
+# XXX Make sure no fds are being leaked *even* if we're keeping all
+#     the relevant objects alive. We shouldn't be depending on reference-count
+#     based cleanup to reclaim fds.
+# XXX Can we double check that _all_ sockets created are being closed
+#     explicitly, rather than as a result of refcounting?
+# XXX Is anyone actually messing with the displayhook? Check!
+# XXX By the time that the Shell stores the sys state (InteractiveShell.save_sys_module_state),
+#     it's already been swapped out for something else. Why? Who's setting the sys
+#     attributes *before* save_sys_module_state gets called, and why?
+# XXX Check list of *all* objects for anything suspiciously IPython-like.
+#     What's the delta between the zeroth run and the first?
+# XXX Consider avoiding the "instance" singleton stuff. Maybe save that
+#     for the rewrite.
+
+
+if six.PY2:
+    def atexit_unregister(func):
+        # Replace the contents, not the list itself, in case anyone else
+        # is keeping references to it.
+        atexit._exithandlers[:] = list(
+            handler for handler in atexit._exithandlers
+            if not handler[0] == func
```

Comment:
Why this instead of `if handler[0] != func`? Something weird about this particular comparison?

Reply:
This is me being defensive: I'm deliberately trying to match the Python 3 C code as closely as possible, and that code uses an `__eq__` comparison rather than a `__ne__` comparison. It's not a given that the two are logical negations of each other: a class can override them both in incompatible ways, or in Python 2 (much more likely) override `__eq__` and neglect to provide the corresponding override for `__ne__`.

Reply:
> that code uses an `__eq__` comparison rather than a `__ne__` comparison

Code link: https://github.com/python/cpython/blob/2fc1160a80733f4c5c88796319154b3f59e98e4b/Modules/atexitmodule.c#L284

Reply:
OK, I figured there was a reason. You should probably leave a note for future readers.

Reply:
Done.